sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,362 @@
+    strategy.
+
+    References
+    ----------
+    .. [1] B. C. Ross "Mutual Information between Discrete and Continuous
+       Data Sets". PLoS ONE 9(2), 2014.
+    """
+    c = c.reshape((-1, 1))
+    n_samples = c.size
+
+    nn = NearestNeighbors(n_neighbors=n_neighbors)
+
+    radius = np.empty(n_samples)
+    label_counts = np.empty(n_samples)
+    for label in np.unique(d):
+        mask = d == label
+        count = np.sum(mask)
+        if count > 1:
+            nn.set_params(n_neighbors=min(n_neighbors, count - 1))
+            nn.fit(c[mask])
+            r = nn.kneighbors()[0]
+            radius[mask] = np.nextafter(r[:, -1], 0)
+        else:
+            radius[mask] = 0
+        label_counts[mask] = count
+
+    nn.set_params(algorithm='kd_tree')
+    nn.fit(c)
+    ind = nn.radius_neighbors(radius=radius, return_distance=False)
+    neighbor_counts = np.array([i.size for i in ind])
+
+    mi = (digamma(n_samples) + digamma(n_neighbors) -
+          np.mean(digamma(label_counts)) -
+          np.mean(digamma(neighbor_counts + 1)))
+
+    return max(0, mi)
+
+
+def _compute_mi(x, y, x_categorical, y_categorical, n_neighbors=3):
+    """Compute mutual information between two variables.
+
+    This is a simple wrapper which selects a proper function to call based on
+    whether `x` and `y` are categorical or not.
+    """
+    if x_categorical and y_categorical:
+        return mutual_info_score(x, y)
+    elif x_categorical and not y_categorical:
+        return _compute_mi_cd(y, x, n_neighbors)
+    elif not x_categorical and y_categorical:
+        return _compute_mi_cd(x, y, n_neighbors)
+    else:
+        return _compute_mi_cc(x, y, n_neighbors)
+
+
+def _get_column(X, i):
+    """Get column of a matrix.
+
+    Parameters
+    ----------
+    X : ndarray or csc_matrix, shape (n_samples, n_features)
+        Matrix from which to get a column.
+    i : int
+        Column index.
+
+    Returns
+    -------
+    xi : ndarray, shape (n_samples,)
+        i-th column of `X` in dense format.
+    """
+    if issparse(X):
+        x = np.zeros(X.shape[0])
+        start_ptr, end_ptr = X.indptr[i], X.indptr[i + 1]
+        x[X.indices[start_ptr:end_ptr]] = X.data[start_ptr:end_ptr]
+    else:
+        x = X[:, i]
+
+    return x
+
+
+def _iterate_columns(X, columns=None):
+    """Iterate over columns of a matrix.
+
+    Parameters
+    ----------
+    X : ndarray or csc_matrix, shape (n_samples, n_features)
+        Matrix over which to iterate.
+    columns : iterable or None, default None
+        Indices of columns to iterate over. If None, iterate over all columns.
+
+    Yields
+    ------
+    xi : ndarray, shape (n_samples,)
+        Columns of `X` in dense format.
+    """
+    if columns is None:
+        columns = range(X.shape[1])
+
+    for i in columns:
+        yield _get_column(X, i)
+
+
+class MutualInfoSelector(BaseEstimator, SelectorMixin):
```

Comment:
@agramfort  The reason I would oppose it to being a function with the default setting is because this algorithm does not take into account just the MI between every feature and the target variable as the name suggest now but also the MI between each of the features.

The algo crudely is something like this,

    1. scores = [MI(feature, target) for features in features]
    2, features_list = [feature corresponding to maximum score]
        for _ in range(n_features_to_select):
            features_list.append( max(scores[feature]) - MI(feature, every_feature_in features_list))

Currently we compute only the necessary `n_features_to_select` highest scores and don't care about the rest. If you want this as a function to be passed into SelectKBest etc.. then it becomes computationally expensive as @nmayorov says . So I would keep it as it is.

Or is there something else you did not understand?